// image-js/lib/browser/environment.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.Canvas = Canvas;
exports.fetchBinary = fetchBinary;
exports.createWriteStream = createWriteStream;
exports.writeFile = writeFile;
var env = 'browser';
var ImageData = self.ImageData;
var DOMImage = self.Image;

function Canvas(width, height) {
    var canvas = self.document.createElement('canvas');
    canvas.width = width;
    canvas.height = height;
    return canvas;
}

function fetchBinary(url, { withCredentials = false } = {}) {
    return new Promise(function (resolve, reject) {
        var xhr = new self.XMLHttpRequest();
        xhr.open('GET', url, true);
        xhr.responseType = 'arraybuffer';
        xhr.withCredentials = withCredentials;

        xhr.onload = function (e) {
            if (this.status !== 200) reject(e);else resolve(this.response);
        };
        xhr.onerror = reject;
        xhr.send();
    });
}

function createWriteStream() {
    throw new Error('createWriteStream does not exist in the browser');
}

function writeFile() {
    throw new Error('writeFile does not exist in the browser');
}

exports.env = env;
exports.ImageData = ImageData;
exports.DOMImage = DOMImage;
